### ISSUE TITLE
[Feature Request] Configurable Value for View mode Grid/List

### ISSUE THREAD

**Issue by xekon** [ampache/ampache#2796]
## Describe what you don't like about Ampache, or what it is missing

If a new user registers, or if you clear your browser cache, then this is the view mode:

![2021-03-04_16-16](https://user-images.githubusercontent.com/106644/110048578-f6452580-7d04-11eb-9523-f47ae49bae3d.png)

I would like to be able to configure Ampache so that the default view mode for new users or after a browser cache clear is this:

![2021-03-04_16-17](https://user-images.githubusercontent.com/106644/110048630-12e15d80-7d05-11eb-9707-05f9862bbd2a.png)

## Describe how you would like the feature to work**

It would be nice if I could configure the default view in the admin panel, while allowing users to change their personal view if they wish.

I would also setting for knowing what file I could edit to change the default view if this is not a feature anyone wants to add.

**Comment by SePHTaN** [comment]
Addition to xekons feature request:
The views for songs, albums interprets are always the same, depending on how you set them in the ui.
So i personally like following view settings:
songs   : listview    (better overview)
albums : gridview  (looks more stylish)
artists   : gridview  (looks more stylish)
artist    : listview of albums
             listview of songs when clicked an album
The view of albums and songs when browsing an artist is the same as the songs view or albums view.
At the moment this can't be achieved because the views can't be set individually for songs, albums, artists and artist.

**Comment by xekon** [comment]
I agree with your personal view likes SePHTan, those are all great defaults.

Also I see what you mean about the two sharing the same setting.
If you click into an individual artist, and change it to list mode, then it also changes the view mode for albums.

**Comment by lachlan-00** [comment]
seems better than the cookie setup used now. it'd prefer the option just to set and forget too

**Comment by xekon** [comment]
Awesome, let me know if you need somebody to test.

**Comment by xekon** [comment]
I have been trying to make the edit to set this. I thought I found the spot in the code to do it, but no luck yet.

I modified `src/Repository/Model/Browse.php` from this:

```
        $name = 'browse_' . $type . '_grid_view';
        if ((filter_has_var(INPUT_COOKIE, $name))) {
            $this->set_grid_view(filter_input(INPUT_COOKIE, $name, FILTER_SANITIZE_STRING,
                    FILTER_FLAG_NO_ENCODE_QUOTES) == 'true');
        }
```

to this:

```
        $name = 'browse_' . $type . '_grid_view';
        if ($type == 'album' OR $type == 'artist' OR $type == 'podcast') {
            if ((filter_has_var(INPUT_COOKIE, $name))) {
                $this->set_grid_view(filter_input(INPUT_COOKIE, $name, FILTER_SANITIZE_STRING,
                        FILTER_FLAG_NO_ENCODE_QUOTES) == 'false');
            }
        } else {
            if ((filter_has_var(INPUT_COOKIE, $name))) {
                $this->set_grid_view(filter_input(INPUT_COOKIE, $name, FILTER_SANITIZE_STRING,
                        FILTER_FLAG_NO_ENCODE_QUOTES) == 'true');
            }
        }
```

The goal was that if it was type album, artist, or podcast, that it would use grid_view instead of list view. I am still poking around making changes here and there but so far without getting the desired outcome.

**Comment by xekon** [comment]
just now I found something else: `src/Module/System/InstallationHelper.php`

```
                setcookie('browse_album_grid_view', 'false', $cookie_options);
                setcookie('browse_artist_grid_view', 'false', $cookie_options);
```

now I just need to figure out where in the code I can have it check for the presence of these cookies, and if they are missing then run the above lines.

**Comment by xekon** [comment]
the only problem with the above solution, is that one the off change somebody wants to change album, artist, or podcast bask to list view, they wont be able to because the cookie always gets over written. I think I need the above code to check if the cookie already exists before issuing setcookie.

**Comment by lachlan-00** [comment]
> the only problem with the above solution, is that on the off chance somebody wants to change album, artist, or podcast bask to list view, they wont be able to because the cookie always gets over written. I think I need the above code to check if the cookie already exists before issuing `setcookie()`, just not sure how to do that yet, I played around with `isset()` but it didn't seem to work.

you can check for cookies with has var, that might help
```
filter_has_var(INPUT_COOKIE, 'browse_album_grid_view')
```

**Comment by xekon** [comment]
I tried filter_has_var and it did not work either, just like isset, I figured out the reason however, basically two copies of the same cookie token were being set, one with hostonly and one without.

I fixed it and found a more appropriate place to put it.

`nano /var/www/ampache/src/Config/Init/InitializationHandlerConfig.php`

```
            // Variables needed for Auth class
            $results['cookie_path']   = $results['raw_web_path'];
            $results['cookie_domain'] = $results['http_host'];
            $results['cookie_life']   = $results['session_cookielife'];
            $results['cookie_secure'] = $results['session_cookiesecure'];

            //make album, artist, and podcast grid view by default, but still changeable by the end user
            $cookie_options = [
                'expires' => time() + 31536000,
                'path' => $results['cookie_path'],
                'domain' => $results['cookie_domain'],
                'secure' => make_bool($results['cookie_secure']),
                'samesite' => 'Strict'
            ];
            if (!isset($_COOKIE['browse_album_grid_view'])) {
                setcookie('browse_album_grid_view', 'false', $cookie_options);
            }
            if (!isset($_COOKIE['browse_artist_grid_view'])) {
                setcookie('browse_artist_grid_view', 'false', $cookie_options);
            }
            if (!isset($_COOKIE['browse_podcast_grid_view'])) {
                setcookie('browse_podcast_grid_view', 'false', $cookie_options);
            }

        }

        // Make sure all default preferences are set
```

**Comment by Jieiku** [comment]
Would still love this feature, I worked out a way to do this programmatically so that I would not have to use nano every time I update/git pull.

```bash
cat << EOF | sudo tee ./src/Config/Init/gridview.php
<?php
    //make album, artist, and podcast grid view by default, but still changeable by the end user
    \$cookie_options = [
        'expires' => time() + 31536000,
        'path' => \$results['cookie_path'],
        'domain' => \$results['cookie_domain'],
        'secure' => make_bool(\$results['cookie_secure']),
        'samesite' => 'Strict'
    ];
    if (!isset(\$_COOKIE['browse_album_grid_view'])) {
        setcookie('browse_album_grid_view', 'false', \$cookie_options);
    }
    if (!isset(\$_COOKIE['browse_artist_grid_view'])) {
        setcookie('browse_artist_grid_view', 'false', \$cookie_options);
    }
    if (!isset(\$_COOKIE['browse_podcast_grid_view'])) {
        setcookie('browse_podcast_grid_view', 'false', \$cookie_options);
    }
?>
EOF

sudo sed -i "/session_cookiesecure/,+2 { s/\}/ include \'gridview\.php\'\;\}/g; }" /var/www/ampache/src/Config/Init/InitializationHandlerConfig.php
```

**Comment by lachlan-00** [comment]
i just found that the cookies were wildly ignored for this. i'll be able to make a preference for it which will set hte cookie on load as now the cookie is checked before the browse

**Comment by lachlan-00** [comment]
this is set after d41fa782c9106c260bb9982d78549b125fe14d4d and 5c2a3122c2c742f804171828e8fe6742005b631b

Now the preference (if set) will overwrite your cookie for the media when you log in. you can change the cookie during your session but it will reset on login. if the preference is off it will ignore any existing cookie

This has been painful but i've been meaning to fix this for a long time. this is one of the ampache if false; then true features that has been really annoying to process.
